### ISSUE TITLE
A new Fonda scripting/launching approach

### ISSUE THREAD

**Issue by kamyshova** [epam/fonda#162]
## Motivation

There are two crucial drawbacks in the current Fonda implementation:

- deadlocks
- idle resources

Both problems are related to the implementation of the scripts launching. There are two types of scripts execution: 

1. The first scripts part (e.g. alignment/post alignment scripts) is launched Fonda at the same time. 
2. The second scripts part (e.g. featureCount, cufflinks) is launched from the alignment/post alignment scripts. 

The work coordination of all scripts is carried out by checking the log files that scripts produce. But log file can not be created at all if the script was not been invoked (e.g. script is launched from the alignment scripts). In this case, post-process scripts will work forever.

![image](https://user-images.githubusercontent.com/23356251/86462320-1a92d180-bd34-11ea-9df4-25f5aeaece87.png)

**Picture 1. The current Fonda launching approach**

> For example, post-process scripts (_qcsummary.sh, cufflinks_cohort.sh_ - see **Picture 1**) are launched with 
> alignment scripts simultaneously. Post-process *cufflinks_cohort.sh* scripts expect the result of the _cufflinks.sh_ 
> script execution by check the _cufflinks.log_ file. But _alignment_ script can fail before _cufflinks.sh_ script invocation. 
> But _cufflinks_cohort.sh_ will not know about it and will run infinitely.

**Deadlocks** are specific for launch in the SGE cluster. Each script is a SGE cluster submitted job. 
The job has specific resource requirements - the number of slots defined by the user in the Fonda global config file 
(_NUMTHREADS_ parameter in _Queue_Parameters_ section). The number of slots is equal to the number of processors in a cluster. 
The user can set such a number of slots that the cluster size will not be enough for job work. In this case, the job hangs on in a pending state (qw). 

> For example, the cluster size is 8 CPU. A user sets _NUMTHREADS=4_.  First of all Fonda launches 3 scripts - _alignment.sh_, _qcsummary.sh_, _cufflinks_cohort.sh_. 2 of them (_alignment.sh_, _qcsummary.sh_) will be in running status. _cufflinks_cohort.sh_ job is in the qw state which stands for being queued and waiting. In its turn, _alignment.sh_ script invokes _cufflinks.sh_ and _featureCount.sh_ and waits for the results. But the cluster doesn't have available slots. _cufflinks.sh_ and _featureCount.sh_ hang on in a pending state, and _alignment.sh_ job will wait for their result endlessly.

Thus, in the beginning post-process jobs take up resources without performing useful work. On the contrary, idle resources are possible in the case of the autoscale cluster. 

## Approach

We propose a new approach to scripts launching.

![image](https://user-images.githubusercontent.com/23356251/86462428-4910ac80-bd34-11ea-84ef-873fa3822cc0.png)

**Picture 2. The new proposed approach**

As can be seen at the picture above, we create an additional **master.sh** script-orchestrator to manage all scripts. 
Fonda will only run **master.sh** script directly. Initially, the master script starts all _alignment.sh_ scripts and waits for their results. After successful completion of alignment step the _cufflinks.sh_, _featureCount.sh_ etc scripts are launched if they are needed.
Please note that we intend to remove the launch of the script from the alignment/post alignment scripts. 
After the per samples scripts are executed successfully, **master.sh** script launches the post-process scripts.

To sum up proposed changes:

- create a new **master.sh** script to manage all scripts
- remove launching of scripts from the alignment/post alignment scripts
- sequential launching of pipeline stages

This approach proposes getting rid of the above problems and makes the process of launching scripts more transparent.
At the same time, this approach preserves the parallelization of processes where it is possible.

**Comment by mgmagi-sanofi** [comment]
I agree with your assessment and am using the idea of a “master controller” in my own dealing with SGE type environment.

I would question though the implementation of the functionality of “master.sh” using Linux shell script.

Higher level programming languages were invented for more than one reason and what you describe is a good fit to be implemented using Java, C++, Python, or others.

I put Java the first only because it would be my personal choice and because essentially Fonda uses (or at least used in the past) the same idea of “master controller” already in the sense that “central” java written code generates and executes per-sample and accumulating shell scripts.

I know the history behind why Java was chosen for that role in Fonda and can assure it was rather incidental, so my arguments are not “pro” Java, but rather “against“ shell script in favor of a more scalable higher level software development platform.

Mark G Magid
Data Management / Software Solutions
Precision Oncology
Desk:      +1 617 665-4295
Mobile:  +1 781 929-9673
Email:     Mark-EXT.Magid@sanofi.com

**Comment by syansanofi** [comment]
I am in favor of this plan. It will be extremely useful for two cases: logging and collaboration.

Currently, FONDA does not have master log functionality, making it difficult to decipher pipeline logic. If orchestration was implemented in Java, then I still do not see how logging can be achieved with ease. The master.sh file should contain an implicit roadmap of the execution logic within it. This change would bring FONDA more in line with other pipelining libraries and languages while preserving OOP advantages of FONDA.

In addition, it will be much easier to share pipeline designs and results with those unfamiliar with Java since shell scripting is more widespread.
